subprojects/core/src/integTest/groovy/org/gradle/groovy/scripts/StaticCompilationTest.groovy

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.groovy.scripts
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+class StaticCompilationTest extends AbstractIntegrationSpec {
+    def "static compilation uses information provided @DelegatesTo and @ClosureArgs in Project"() {
+        /*
+        This test contains many (seemingly meaningless) assignments such as
+        'Foo x = delegate'.  These serve two purposes.
+        1) If, in the above example, the type 'Foo' did not match the type 'delegate',
+           Groovy's static compiler would reject it as invalid.  This verifies the type
+           information provided by closures annotated with @DelegatesTo and @ClosureArgs
+           is being used by static compilation.
+        2) If, in the above example, 'it' was not _actually_ of type 'Foo', Groovy's run-
+           time would reject it at execution time as invalid.  This helps verify the type
+           information provided by closures annotated with @DelegatesTo and @ClosureArgs
+           is actually _correct_.
+        */
+        // TODO? How to validate @DelegatesTo strategy?
```

Comment:
I couldn't figure out a way to validate ```@DelegatesTo``` with this approach, unfortunately, nor how to do any kind of meaningful check for static compilation failures.  But, you can experiment with modifying the ```@DelegatesTo``` and ```@ClosureParams``` in Project, and watch the effects on this test.